### PR TITLE
New GUI

### LINKED ISSUE #28: [Feature Request]: Design proposal
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Potential Enhancement

The current UI isn't the best and [a rewrite is planned](https://github.com/Fabulously-Optimized/vanilla-installer/blob/7447eb7b423e64fff896fc4cee657e1e7cafde3d/docs/planned-features.md) anyways
I have thrown together a design and I'd like to see what you think of it

### How it could be implemented

light mode:

dark mode:

Figma link: https://www.figma.com/file/KmhApEpUavYCJBSqBUnXbh/FO-Vanilla-installer

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by tuxinal**
Implements the new GUI proposed in #28 

image

I haven't tested this on Windows and other python versions besides 3.10 yet. please do so if you have the chance
_(you might also have to install the [Inter](https://rsms.me/inter/) font until i figure out how to include it in the application itself)_

### CODE REVIEW

**Comment by tuxinal** [comment]
@nsde i have made some changes to how the `get_dir`/`set_dir` functions work in `main.py`. the current implementation made the packwiz bootstrapper appear inside the home directory on linux systems and it seems unnecessary as far as i can tell. i have discussed this with osfanbuff on discord and they said it should be okay. i wanted to ask if you think there is any reason to keep the current functionality.

**Comment by nsde** [comment]
LGTM, but:
- [Accessing the API may be problematic because it can lead to ratelimiting](https://github.com/Fabulously-Optimized/vanilla-installer/compare/main...tuxinal:vanilla-installer:main#diff-253c6a5418c0bf267e44dfb8f97def5223ece1268f602d15ca8813b76313d410R283)
- The theme dictionary keys have in-descriptive names, such as `subtext1` and `subtext0`

Also, as of right now, the fork don't actually implement the new GUI, but the commits look like preparations for that.

**Comment by osfanbuff63** [comment]
> Also, as of right now, the fork don't actually implement the new GUI, but the commits look like preparations for that

Tested it briefly locally, and it does seems to implement it. Haven't gone through everything yet, so I'll leave a review once I get through it.

Only thing I noticed is that it doesn't work on 3.11 but that's pyside6's fault, we'll just need to wait there.

**Comment by tuxinal** [comment]
the theme names are based on [catppuccin](https://github.com/catppuccin/catppuccin)'s names for the colors. i suppose i can change them to be more descriptive.
what method would you suggest to avoid rate limits?

File: vanilla-installer/main.py [hunk 1]
```diff
@@ -318,10 +273,28 @@ def create_profile(mc_dir: str, version_id: str) -> None:
...
     launcher_profiles_path.write_text(profiles_json)
 
+def get_pack_mc_versions() -> list[str]:
+    """Gets a list of all the versions FO currently supports
+    """
+    exp = re.compile(r'\d+\.\d+(\.\d+)?')
+    return_value = []
+    try:
+        response = requests.get(
+            "https://api.github.com/repos/Fabulously-Optimized/fabulously-optimized/contents/Packwiz").json()
```
Comment (osfanbuff63) [inline vanilla-installer/main.py]: I think probably the easiest way to combat ratelimiting (which I believe is IP-based when unauthenticated, so it *shouldn't* cause other users trouble) is probably to cache the list and not make the request if the cache is found - cache basically meaning opening a text file and storing it there, although there's surely a better way to do that.

File: vanilla-installer/theme.py [hunk 2]
```diff
@@ -40,37 +40,48 @@ def is_dark(to_dark: bool = None) -> bool:
         return open(FILE, "w", encoding="utf-8").write("dark")
     return open(FILE, encoding="utf-8").read() == "dark"
 
+# colors from catppuccin latte and mocha https://github.com/catppuccin/catppuccin
```
Comment (osfanbuff63) [inline vanilla-installer/theme.py:43]: Probably worth adding a mention in credits.md for this one. Speaking of credits, is it worth adding a button for that?

File: vanilla-installer/gui.py [hunk 3]
```diff
@@ -22,185 +35,234 @@
...
+    ui = Ui_MainWindow()
+    ui.setupUi(window)
+    get_versions_worker = Worker(ui.addVersions)
+    ui.threadpool.start(get_versions_worker)
+    window.show()
+    app.exec()
+class Ui_MainWindow(object):
+    def setupUi(self, MainWindow: QMainWindow):
+        if not MainWindow.objectName():
+            MainWindow.setObjectName("MainWindow")
```
Comment (osfanbuff63) [inline vanilla-installer/gui.py:46]: Seems at least on Windows, the window name is taken from here - so should probably be `VanillaInstaller` or something similar.

**Review by osfanbuff63 — COMMENTED** [review]
As mentioned earlier, Qt doesnt have a build out yet for python 3.11 (although likely all that's needed is a dep update) which isn't your fault ofc - so it doesn't work on 3.11 yet.

A couple bugs here and there, but overall it's pretty good (and works)! It might be worth adding a `QIcon`, and it doesn't scale when you maximize it, but those are relatively trivial.

File: vanilla-installer/theme.py (hunk 2 above)
Comment (tuxinal) [reply vanilla-installer/theme.py:43]: The icons are also from feathericons. we can add that to credit.md too

File: vanilla-installer/main.py (hunk 1 above)
Comment (tuxinal) [reply vanilla-installer/main.py]: We can also add a settings menu with an option to add a github api key, i doubt anyone who's gonna install FO more than 60 times in an hour doesn't have a github account

File: vanilla-installer/gui.py (hunk 3 above)
Comment (tuxinal) [reply vanilla-installer/gui.py:46]: can you try the new commit and see if the title changed?

Comment (osfanbuff63) [reply vanilla-installer/gui.py:46]: Once I'm home I'll try it yeah

Comment (osfanbuff63) [reply vanilla-installer/gui.py:46]: Yep, works now

File: vanilla-installer/gui.py [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
+    QTextEdit,
+    QWidget,
+    QGraphicsColorizeEffect,
+    QFileDialog
+)
+from PySide6.QtSvgWidgets import QSvgWidget
+import minecraft_launcher_lib as mll
 
 # External
 import requests
```
Comment (osfanbuff63) [inline vanilla-installer/gui.py]:
```suggestion

```
unused import, not a huge deal though
(same with logger but I can't leave a review comment there)

File: vanilla-installer/gui.py [hunk 5]
```diff
@@ -22,185 +32,234 @@
 parser.add_argument("--safegui", type=bool)
 parser.add_argument("--litegui", type=bool)
```
Comment (osfanbuff63) [inline vanilla-installer/gui.py]: These are unused now, the GUI isn't meant to be invoked via a CLI anyway since we have a dedicated CLI.

**Review by osfanbuff63 — COMMENTED** [review]
Couple things, but otherwise looking pretty good. If you (or anyone else) wants to do a PR for a settings menu for an API token or something like that, feel free - but I'd kinda like to get this merged if possible.

File: vanilla-installer/main.py (hunk 1 above)
Comment (osfanbuff63) [reply vanilla-installer/main.py]: If you or anyone else wants to do a separate PR for this that's fine, but I'd kinda like to get this merged 😅

File: vanilla-installer/main.py [hunk 6]
```diff
@@ -81,14 +65,48 @@ def get_dir() -> str:
         path = set_dir(default_dir)
     return path
 
+def set_gh_auth(user: str, key: str) -> bool | None:
+    """Sets the GitHub authentication details to be used by the GitHub api.
+    Args:
+        key (str): new username
```
Comment (osfanbuff63) [inline vanilla-installer/main.py]:
```suggestion
        user (str): new username
```

**Review by nsde — APPROVED** [review]

**Review by osfanbuff63 — APPROVED** [review]
Looks good!

**Comment by osfanbuff63** [comment]
@all-contributors add @tuxinal for code

**Pull request merged**